File: CPractice/merge sort.c
#include<stdio.h>
void merge(int [],int,int,int);
void mergesort(int [],int,int);
void merge(int a[],int beg,int mid,int end)
{
    int i,j,k,n1,n2;
    n1=mid-beg+1;
    n2=end-mid;
    int a1[n1],a2[n2];
    for(i=0;i<n1;i++)
        a1[i]=a[beg+i];
    for(j=0;j<n2;j++)
        a2[j]=a[mid+1+j];
    i=0,j=0,k=beg;
    while(i<n1 && j<n2)
    {
        if(a1[i]<=a2[j])
        {
            a[k]=a1[i];
            i++;
        }
        else
        {
            a[k]=a2[j];
            j++;
        }
        k++;
    }
    while(i<n1)
    {
        a[k]=a1[i];
        i++;
        k++;
    }
    while(j<n2)
    {
        a[k]=a2[j];
        j++;
        k++;
    }
}
void mergesort(int a[],int beg,int end)
{
    if(beg<end)
    {
        int mid=(beg+end)/2;
        mergesort(a,beg,mid);
        mergesort(a,mid+1,end);
        merge(a,beg,mid,end);
    }
}
int main()
{
    int n,a[100],i,j,p;
    scanf("%d",&n);
    for(i=0;i<n;i++)
    {
        scanf("%d",&a[i]);
    }
    mergesort(a,0,n-1);
    for(i=0;i<n;i++)
    {
        printf("%d ",a[i]);
    }
    printf("\n"); 
}

File: DSA codes/Recursion and Backtracking/merge_sort.py
class Solution: # Divide and merge
    def mergeSort(self,arr, low, high):
        if low>=high:
            return
        mid = (low + high)//2
        self.mergeSort(arr, low, mid)
        self.mergeSort(arr, mid+1, high)

        left = low
        right = mid+1   

        temp_array = []
        while left <= mid and right <= high:
            if arr[left] < arr[right]:
                temp_array.append(arr[left])
                left+=1
            else:
                temp_array.append(arr[right])
                right+=1
            
        while left <= mid:
            temp_array.append(arr[left])
            left+=1
        while right <= high:
            temp_array.append(arr[right])
            right+=1
        
        for i in range(low,high+1):
            arr[i] = temp_array[i-low]


if __name__ == "__main__":
    obj = Solution()
    arr = [4,6,3,2,8,7,1]
    obj.mergeSort(arr,0, len(arr)-1)
    print(arr)